### PR TITLE
Migrate state_notifier to null-safety

### PR DESCRIPTION
**Pull request by gaetschwartz**
Migrated state_notifier to null-safety, tests are not migrated yet because mockito hasn't migrated yet. They all pass though !

### CODE REVIEW

**Comment by gaetschwartz** [comment]
CI is failing because it's only on Beta and Dev hence we'd need to publish it as a prerelease. I'm unsure about how to version this so I put `0.7.0-nullsafety` as it seems to be what others do ?

**Comment by rrousselGit** [comment]
Indeed, we'll have to do that.

**Comment by gaetschwartz** [comment]
Coverage seems to fail on Beta for some reason. I don't know test_coverage enough to say more

**Comment by rrousselGit** [comment]
The tests are failing I think. See the CI on master – that's probably what is happening here.

**Comment by gaetschwartz** [comment]
It is test_coverage that is failing, the tests pass. Why not use `dart test --coverage coverage` instead ? Saves the coverage file as a `.json` in `coverage/test/`.
image

**Comment by rrousselGit** [comment]
> Why not use dart test --coverage coverage instead ? Saves the coverage file as a .json in coverage/test/.

Codecov requires an lcov file, which is not generated by `dart test --coverage`. This scripts generates one

We can remove `test_coverage`, but we need to generate an lcov file.

**Comment by gaetschwartz** [comment]
`flutter test --coverage` does though. Strange that dart doesn't. Maybe we can use flutter for now ?

**Comment by rrousselGit** [comment]
We can't.
`flutter test` does not work on Dart-only packages, and state_notifier is one

**Comment by gaetschwartz** [comment]
It does on my machine at least, just have to add 

```
  flutter_test:
    sdk: flutter
```

to the `pubspec.yaml` and it does, unless you specifically want to use dart test it might be an alternative. But I don't really understand why is test_coverage failing. It's probably not compatible with NNBD I guess?

image

**Comment by rrousselGit** [comment]
We can't add flutter_test as dependency. The package must not depend on flutter

**Comment by gaetschwartz** [comment]
Hum okay so apparently `dart test` and `dart .\test\state_notifier_test.dart` don't do the same things, dart test works but the latter fails, will investigate more and try to fix these.

**Comment by gaetschwartz** [comment]
Indeed test_coverage uses `dart .\test\state_notifier_test.dart` which also fails on `master`. I don't really know what that implies, are the tests actually failing and dart test is inaccurate or the opposite and using `dart .\test\state_notifier_test.dart`  doesn't make any sense ?

image

**Comment by rrousselGit** [comment]
You need to make sure that you passed the `--enable-asserts` flag, otherwise, the code behaves differently

**Comment by gaetschwartz** [comment]
Alright so the tests indeed pass, I think the issue is that `test_coverage` generates a file that doesn't start with

**Comment by gaetschwartz** [comment]
Alright this should fix CI, instead of using `test_coverage`, we use `coverage`. The package translates the `.json` coverage file to to a `lcov` one.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -42,11 +42,11 @@ jobs:
         working-directory: ${{ matrix.package }}
 
       - name: Run tests
-        run: dart test
+        run: dart test --coverage coverage
         working-directory: ${{ matrix.package }}
 
       - name: Coverage
-        run: dart pub run test_coverage
+        run: dart run coverage:format_coverage -l -i ./coverage/test/state_notifier_test.dart.vm.json -o ./coverage/lcov.info --packages ./.packages
```
Comment (rrousselGit) [inline .github/workflows/build.yml:49]: That `state_notifier_testdart.vm.json` worries me. Does that means it tracks the coverage only for this file?

Would something like `./test/**/*.vm.json` works?

**Comment by rrousselGit** [comment]
Thanks again! Using `coverage` sounds interesting. I'll play around with it too, to see if that works correctly

File: .github/workflows/build.yml (hunk 1 above)
Comment (gaetschwartz) [reply .github/workflows/build.yml:49]: I don't know how and if `format_coverage `handles multiple files. I'll have to test it out. It's true that it's not the best to have one specific filename. Will investigate.

Comment (rrousselGit) [reply .github/workflows/build.yml:49]: Just tested, it appears that `./coverage/test/**/*.dart.vm.json` works with multiple files

Comment (gaetschwartz) [reply .github/workflows/build.yml:49]: And it combines it into one single lcov file ? If yes that's perfect but make sure it does and doesn't just overwrite it with the latest one

**Pull request merged**